client/jetbrains/src/main/kotlin/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.kt

```diff
@@ -0,0 +1,348 @@
+import com.sourcegraph.cody.autocomplete.render.CodyAutocompleteSingleLineRenderer
+import com.sourcegraph.cody.autocomplete.render.InlayModelUtil.getAllInlaysForEditor
+import com.sourcegraph.cody.statusbar.CodyAutocompleteStatus
+import com.sourcegraph.cody.statusbar.CodyAutocompleteStatusService.Companion.notifyApplication
+import com.sourcegraph.cody.statusbar.CodyAutocompleteStatusService.Companion.resetApplication
+import com.sourcegraph.cody.vscode.CancellationToken
+import com.sourcegraph.cody.vscode.InlineAutocompleteItem
+import com.sourcegraph.cody.vscode.InlineAutocompleteList
+import com.sourcegraph.cody.vscode.InlineCompletionTriggerKind
+import com.sourcegraph.cody.vscode.IntelliJTextDocument
+import com.sourcegraph.cody.vscode.TextDocument
+import com.sourcegraph.config.ConfigUtil.isCodyEnabled
+import com.sourcegraph.config.UserLevelConfig
+import com.sourcegraph.telemetry.GraphQlLogger
+import com.sourcegraph.utils.CodyEditorUtil.getAllOpenEditors
+import com.sourcegraph.utils.CodyEditorUtil.getLanguage
+import com.sourcegraph.utils.CodyEditorUtil.getTextRange
+import com.sourcegraph.utils.CodyEditorUtil.isCommandExcluded
+import com.sourcegraph.utils.CodyEditorUtil.isEditorValidForAutocomplete
+import com.sourcegraph.utils.CodyEditorUtil.isImplicitAutocompleteEnabledForEditor
+import difflib.Delta
+import difflib.DiffUtils
+import difflib.Patch
+import java.util.*
+import java.util.concurrent.CancellationException
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionException
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Consumer
+import java.util.stream.Collectors
+
+/** Responsible for triggering and clearing inline code completions (the autocomplete feature). */
+@Service
+class CodyAutocompleteManager {
+  private val logger = Logger.getInstance(CodyAutocompleteManager::class.java)
+  private val currentJob = AtomicReference(CancellationToken())
+  var currentAutocompleteTelemetry: AutocompleteTelemetry? = null
+
+  /**
+   * Clears any already rendered autocomplete suggestions for the given editor and cancels any
+   * pending ones.
+   *
+   * @param editor the editor to clear autocomplete suggestions for
+   */
+  @RequiresEdt
+  fun clearAutocompleteSuggestions(editor: Editor) {
+    // Log "suggested" event and clear current autocompletion
+    Optional.ofNullable(editor.project).ifPresent { p: Project? ->
+      if (currentAutocompleteTelemetry != null &&
+          currentAutocompleteTelemetry!!.getStatus() !=
+              AutocompletionStatus.TRIGGERED_NOT_DISPLAYED) {
+        currentAutocompleteTelemetry!!.markCompletionHidden()
+        GraphQlLogger.logAutocompleteSuggestedEvent(
+            p!!,
+            currentAutocompleteTelemetry!!.latencyMs,
+            currentAutocompleteTelemetry!!.displayDurationMs,
+            currentAutocompleteTelemetry!!.params())
+        currentAutocompleteTelemetry = null
+      }
+    }
```

Comment:
Completely stylistically, this might read a bit easier than `!!` :smile: (requires `gradle spotlessApply`)

```suggestion
  editor.project?.let { p ->
      currentAutocompleteTelemetry?.let { autocompleteTelemetry ->
        if (autocompleteTelemetry.status != AutocompletionStatus.TRIGGERED_NOT_DISPLAYED) {
          autocompleteTelemetry.markCompletionHidden()
          GraphQlLogger.logAutocompleteSuggestedEvent(
            p,
            autocompleteTelemetry.latencyMs,
            autocompleteTelemetry.displayDurationMs,
            autocompleteTelemetry.params())
          currentAutocompleteTelemetry = null
        }
      }
    }
```

Reply:
done